test/Microsoft.Extensions.Caching.Memory.Tests/CapacityTests.cs

```diff
@@ -0,0 +1,391 @@
+                SizeLimit = 10,
+                CompactionPercentage = 0.5
+            });
+
+            Assert.Equal(0, cache.Size);
+
+            cache.Set("key", "value", new MemoryCacheEntryOptions { Size = 6 });
+
+            Assert.Equal("value", cache.Get("key"));
+            Assert.Equal(6, cache.Size);
+
+            cache.Set("key", "value1", new MemoryCacheEntryOptions { Size = 11 });
+
+            Assert.Null(cache.Get("key"));
+            Assert.Equal(0, cache.Size);
+        }
+
+        [Fact]
+        public void RemovingEntryDecreasesCacheSize()
+        {
+            var cache = new MemoryCache(new MemoryCacheOptions
+            {
+                ExpirationScanFrequency = TimeSpan.Zero,
+                SizeLimit = 10,
+                CompactionPercentage = 0.5
+            });
+
+            cache.Set("key", "value", new MemoryCacheEntryOptions { Size = 5 });
+
+            Assert.Equal(5, cache.Size);
+
+            cache.Remove("key");
+
+            Assert.Equal(0, cache.Size);
+        }
+
+        [Fact]
+        public void ExpiringEntryDecreasesCacheSize()
+        {
+            var cache = new MemoryCache(new MemoryCacheOptions
+            {
+                ExpirationScanFrequency = TimeSpan.Zero,
+                SizeLimit = 10,
+                CompactionPercentage = 0.5
+            });
+
+            var entryOptions = new MemoryCacheEntryOptions { Size = 5, };
+            var changeToken = new TestExpirationToken();
+            entryOptions.ExpirationTokens.Add(changeToken);
+
+            cache.Set("key", "value", entryOptions);
+
+            Assert.Equal(5, cache.Size);
+
+            // Expire entry
+            changeToken.Fire();
+
+            // Trigger compaction
+            Assert.Null(cache.Get("key"));
+
+            // Wait for compaction to complete
+            Thread.Sleep(TimeSpan.FromSeconds(1));
+
+            Assert.Equal(0, cache.Size);
+        }
+
+        [Fact]
+        public void CompactsToLessThanLowWatermarkUsingLRUWhenHighWatermarkExceeded()
+        {
+            var testClock = new TestClock();
+            var cache = new MemoryCache(new MemoryCacheOptions
+            {
+                ExpirationScanFrequency = TimeSpan.Zero,
+                Clock = testClock,
+                SizeLimit = 10,
+                CompactionPercentage = 0.3
+            });
+
+            for (var i = 1; i < 5; i++)
+            {
+                cache.Set($"key{i}", $"value{i}", new MemoryCacheEntryOptions { Size = i });
+                testClock.Add(TimeSpan.FromSeconds(1));
+            }
+
+            // There should be 4 items in the cache
+            Assert.Equal(4, cache.Count);
+
+            cache.Set("key5", "value5", new MemoryCacheEntryOptions { Size = 1 });
+            testClock.Add(TimeSpan.FromSeconds(10));
+
+            // Wait for compaction to complete
+            Thread.Sleep(TimeSpan.FromSeconds(4));
```

Comment:
Increased these waits significantly to ensure tests pass on appveyor and travis. This is a rather poor approach and I'll see if there are better ways to ensure the consistency of these tests.